src/channel-qvm.lisp

```diff
@@ -0,0 +1,374 @@
+
+
+(defun matches-rule (rule instr-qubits instr-name position)
+  "check if rule is matched by instruction data and the position of the match request"
+  (let* ((noise-predicate (predicate (noise-predicate rule)))
+         (noise-position (noise-position (noise-predicate rule))))
+    (and (funcall noise-predicate instr-qubits instr-name) (string= position noise-position))))
+
+
+(defun match-rules (rules instr-qubits instr-name position)
+  "return the operation elements for the first matching rule"
+  (loop :for rule in rules
+        :when (matches-rule rule instr-qubits instr-name position)
+          :return (operation-elements rule)))
+
+
+(defmethod apply-all-kraus-operators ((qvm channel-qvm) (instr quil:gate-application) kraus-operators)
+  "Apply all the kraus operators to the state of the system."
+  (loop :for kops :in kraus-operators
+        :do
+           (check-kraus-ops kops)
+           (apply-kraus-ops qvm instr kops)))
+
+
+(defmethod apply-kraus-ops ((qvm channel-qvm) (instr quil:gate-application) kraus-ops)
+  (let* ((amps (%trial-amplitudes qvm))
+         (r (random 1.0d0))
+         (summed-p 0.0d0)
+         (logical-qubits (quil:application-arguments instr))
+         (qubits (mapcar #'quil:qubit-index logical-qubits))) ;; this needs to be some kind of reference to operational qubits. Have another dict? 
+    (check-type amps quantum-state)
+    (loop :for kj :in kraus-ops
+          :do
+             (replace amps (amplitudes qvm))
+             (apply-matrix-operator (magicl-matrix-to-quantum-operator kj) amps (apply #'nat-tuple qubits))
+             (incf summed-p (psum #'probability amps))
+          :until (>= summed-p r))
+
+    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))
+    (normalize-wavefunction (amplitudes qvm))))
+
+
+(defgeneric apply-classical-readout-noisee (qvm instr)
+  (:documentation "Given a QVM and a (measurement) instruction INSTR, corrupt the readout bit according to the POVM specifications of QVM.")
+  ;; Pure state QVM has no readout noise.
+
+  (:method ((qvm pure-state-qvm) (instr quil:measurement))
+    (declare (ignore qvm instr))
+    nil)
+  ;; Readout noise only happens to the resulting classical bit (i.e.,
+  ;; it's classical noise). As such, discarding that bit doesn't
+  ;; warrant any sort of special treatment.
+  (:method ((qvm approx-qvm) (instr quil:measure))
+    (%corrupt-readout qvm instr (readout-povms qvm))
+    )
+  (:method ((qvm channel-qvm) (instr quil:measure-discard))
+    (declare (ignore qvm instr))
+    nil)
+  ;; We do have a readout bit, and we want to corrupt it.
+  (:method ((qvm channel-qvm) (instr quil:measure))
+    (%corrupt-readout qvm instr (readout-povms (noise-model qvm))))
+  ;; For compiled measurements, refer to the source of that
+  ;; instruction.
+  (:method ((qvm channel-qvm) (instr compiled-measurement))
+    (apply-classical-readout-noise qvm (source-instruction instr))))
+
+
+(defun %corrupt-readout (qvm instr povm-map)
```

Comment:
IIUC, it looks like this doing the same thing as `%corrupt-qvm-memory-with-povm` in noisy-qvm.lisp, except that it takes the `povm-map` as an argument rather than expecting the `qvm` to have a `readout-povms` method?

My $0.02: I think it would be OK to modify the definition of `%corrupt-qvm-memory-with-povm` in noisy-qvm.lisp to take the `povm-map` as an argument and then just call it rather than having this separate function.

Alternatively, I think you could define a `readout-povms` method for the `channel-qvm` like so (warning untested):

```lisp
(defmethod readout-povms ((qvm channel-qvm))
  (readout-povms (noise-model qvm)))
```

then use `%corrupt-qvm-memory-with-povm` as-is and get rid of `%corrupt-readout`.

Reply:
Yup -- `%corrupt-readout` in `noisy-qvm.lisp` takes `povm-map` as a param whereas `%corrupt-qvm-memory-with-povm` expects the qvm to have a `readout-povms` method.  

I chose to "copy" the noisy-qvm.lisp function here because if these noise models eventually replace noisy-qvm, then I thought it would be better to have this separate function in the `channel-qvm` file. If it would be cleaner for me to just target `corrupt-qvm-memory-with-povm` then I can go ahead and change this.

Reply:
That's a good point. My preference would be for a single function if it can be made to work without too much hassle, but either is fine I guess.

Reply:
Ok, I think a good solution is to just change the `noisy-qvm.lisp` `apply-classical-readout-noise` to just target %corrupt-readout instead of %corrupt-qvm-memory-with-povm. That way if we end up deprecating `noisy-qvm` we don't have to make changes to `channel-qvm.lisp`